feat(filesystem): add reference grammar for the filesystem_info interface (#22)
charmed-hpc/docs: reference/filesystem-charms/index.md

(filesystem-charms)=
# Filesystem Charms

## `filesystem_info`

> The [`filesystem_info`](https://charmhub.io/filesystem-client/libraries/filesystem_info) interface
> uses URIs to share the mount information between the server and the client. This allows using the same
> data format for all filesystem types.

Reviewer: ```suggestion
The [`filesystem_info`](https://charmhub.io/filesystem-client/libraries/filesystem_info) interface
uses URIs to share the mount information between the server and the client. This allows for the use of the same
data format for all filesystem types.
```

The grammar of the specific shape of URIs used is defined as:

:::{code-block}
key = 1*( unreserved )
value = 1*( unreserved / ":" / "/" / "?" / "#" / "[" / "]" / "@" / "!" / "$"
      / "'" / "(" / ")" / "*" / "+" / "," / ";" )
options = key "=" value ["&" options]

host-port = host [":" port]
hosts = host-port ["," hosts]
authority = [userinfo "@"] "(" hosts ")"

URI = scheme "://" authority path-absolute ["?" options]
:::

> Any unspecified grammar rule is specified by [RFC 3986](https://datatracker.ietf.org/doc/html/rfc3986#appendix-A).

Reviewer: Having the RFC link here as well is useful. For users who aren't as familiar with this formatting/grammar, is there an additional link we can provide that provides examples and/or more details?

Author: We can point to the examples listed on the RFC:
https://datatracker.ietf.org/doc/html/rfc3986#section-1.1.2

Reviewer 2: Re. the url rules here, I think we should include some examples of filesystem urls that match here. I don't think that I'm going out on a limb here by saying that most "mere mortals" and HPC system admins probably aren't all that familiar with regular language or RFC 3986; they know how to deploy applications and want to know what the look for in logs when things break.

I think some examples of filesystem urls provided by the filesystem proxy operators could go a long way here with additional linkage to RFC 3986.

### Additional requirements

> - The scheme component must identify the type of filesystem that needs to be mounted by the client.

Reviewer: ```suggestion
- The `scheme` component must identify the type of filesystem that to be mounted by the client.
```

Reviewer: "Such as..." - could we add an example or two?

> - The userinfo component may contain a user for authentication purposes, but it must not contain

Reviewer: ```suggestion
- The `userinfo` component may contain a user for authentication purposes, but it must not contain
```

  any password required to authenticate against the filesystem. In other words, the `user:password` syntax is not allowed.
- The hosts component must contain the host or list of hosts encompassing the server. The characters
  `(` and `)` must encompass an array of values, and the character `,` must delimit each host ip or domain name. If only
  a single host is required, the host must still be encompassed by `(` and `)`.
- The path-absolute component may be the exported path of the filesystem.
- The options component may contain any other required data for the specific filesystem type, including but not limited to:
      - Password to authenticate an user (raw or secret based).
      - Cluster identifier.
      - Filesystem identifier.

> - Since each filesystem type will require different data on its options component, and it is unknown if more data will be
>   required in the future, the scheme component may be used to version different data formats for the same filesystem type,
>   which doesn’t introduce any breaking changes.

Reviewer: ```suggestion
- Since each filesystem type will require different data for its options component, and it is unknown if more data will be
  required in the future, the scheme component may be used to version different data formats for the same filesystem type,
  which will not introduce any breaking changes.
```

> Any filesystem providers must adhere to both the previously defined grammar and the additional requirements specified.

Reviewer: ```suggestion
Filesystem providers must adhere to both the defined grammar and the additional requirements specified here.
```
